### PR TITLE
Add tool stitch

### LINKED ISSUE #32: Include stitch imputation tool
### Description of feature

Add STITCH software as one of the imputation modes.

### LINKED ISSUE #33: Select two tools at the same time
### Description of the bug

The pipeline currently allows for selecting only one single imputation tool. This behavior should be modified as users may want to use more than one tool for comparison.

```
ERROR ~ ERROR: Validation of pipeline parameters failed!

 -- Check '.nextflow.log' file for details
The following invalid input values have been detected:

* --tools: 'glimpse1,quilt' is not a valid choice (Available choices: glimpse1, glimpse2, quilt)
```

### Command used and terminal output

```console
nextflow run phaseimpute -profile test,singularity --outdir test_both --tools glimpse1,quilt
```

### Relevant files

_No response_

### System information

_No response_

### LINKED ISSUE #41: Separate get panel into subworkflows

### PR DESCRIPTION
**Pull request by atrigila**
Closes #32 
Closes #33 
Closes #41

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/phaseimpute/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/phaseimpute _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/nf-core/stitch/stitch.diff [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+--- modules/nf-core/stitch/main.nf
++++ modules/nf-core/stitch/main.nf
+@@ -8,8 +8,8 @@
+         'biocontainers/r-stitch:1.6.10--r43h06b5641_0' }"
+ 
+     input:
+-    tuple val(meta) , path(posfile), path(input, stageAs: "input"), path(rdata, stageAs: "RData_in"), val(chromosome_name), val(K), val(nGen)
+-    tuple val(meta2), path(collected_crams), path(collected_crais), path(cramlist)
++    tuple val(meta2), path(posfile), path(input, stageAs: "input"), path(rdata, stageAs: "RData_in"), val(chromosome_name), val(K), val(nGen)
++    tuple val(meta), path(collected_crams), path(collected_crais), path(cramlist)
```
Comment (LouisLeNezet) [inline modules/nf-core/stitch/stitch.diff:11]: Why this modification ?
Shouldn't the cram channel be at the top ?

File: nextflow_schema.json [hunk 2]
```diff
@@ -73,13 +73,13 @@
                 },
                 "min_val_gl": {
                     "type": "number",
-                    "description": "Minimum genotype likelihood probability P(G|R) in validation data. Set to zero to have no filter of if using –gt-validation",
+                    "description": "Minimum genotype likelihood probability P(G|R) in validation data. Set to zero to have no filter of if using \u2013gt-validation",
```
Comment (LouisLeNezet) [inline nextflow_schema.json:76]: Is there a reason to go for a unicode number for a simple dash ?

File: subworkflows/local/bam_impute_stitch/bam_impute_stitch.nf [hunk 3]
```diff
@@ -0,0 +1,31 @@
+include { STITCH             } from '../../../modules/nf-core/stitch/main'
+include { BCFTOOLS_INDEX     } from '../../../modules/nf-core/bcftools/index/main'
+
+
+workflow BAM_IMPUTE_STITCH {
+
+    take:
+    ch_parameters                                //  channel: [ val(meta), bam, bai ]
```
Comment (LouisLeNezet) [inline subworkflows/local/bam_impute_stitch/bam_impute_stitch.nf:8]: Could you update the channel description to match the input ?

File: nextflow_schema.json [hunk 4]
```diff
@@ -428,6 +428,14 @@
         "buffer": {
             "type": "integer",
             "default": 10000
+        },
+        "k_val": {
+            "type": "integer",
+            "default": 2
+        },
+        "seed": {
```
Comment (LouisLeNezet) [inline nextflow_schema.json:436]: Should this seed be passed also to Quilt ?

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    ch_versions      = Channel.empty()
+
+    // Prepare posfile and parameters for STITCH
+    // Convert position file to tab-separated file
+    BCFTOOLS_QUERY(ch_panel_sites, [], [], [])
+    ch_posfile = BCFTOOLS_QUERY.out.output
+
+    // Remove multiallelic positions from tsv
+    GAWK(ch_posfile, [])
```
Comment (LouisLeNezet) [inline subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:23]: Shouldn't this go in the panel preparation ?

File: subworkflows/local/vcf_concatenate_bcftools/main.nf [hunk 6]
```diff
@@ -12,7 +12,7 @@ workflow VCF_CONCATENATE_BCFTOOLS {
 
     // Remove chromosome from meta
     ch_vcf_tbi_grouped = ch_vcf_tbi
-        .map{ meta, vcf, tbi -> [['id' : meta.id], vcf, tbi] }
+                    .map{ meta, vcf, tbi -> [['id' : meta.id], vcf, tbi] }
```
Comment (LouisLeNezet) [inline subworkflows/local/vcf_concatenate_bcftools/main.nf:15]: Why this extra space ?

File: workflows/phaseimpute/main.nf [hunk 7]
```diff
@@ -100,32 +109,41 @@ workflow PHASEIMPUTE {
...
-        ch_panel_sites_tsv = GET_PANEL.out.panel
-            .map{ metaPC, norm, n_index, sites, s_index, tsv, t_index, phased, p_index
-                -> [metaPC, sites, tsv]
-            }
-        CONCAT_PANEL(GET_PANEL.out.panel
-            .map{ metaPC, norm, n_index, sites, s_index, tsv, t_index, phased, p_index
-                -> [[id:metaPC.panel], sites, s_index]
-            }
+        // Normalize indels in panel
+        VCF_NORMALIZE_BCFTOOLS(VCF_CHR_CHECK.out.vcf, ch_fasta)
```
Comment (LouisLeNezet) [inline workflows/phaseimpute/main.nf:117]: Don't forget to add the versions of the software to the ch_versions channel as well as usefull stats and files to the multiqc channel ? 😉

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf [hunk 8]
```diff
@@ -0,0 +1,71 @@
+include { BCFTOOLS_VIEW as VIEW_VCF_SNPS         } from '../../../modules/nf-core/bcftools/view/main.nf'
+include { BCFTOOLS_VIEW as VIEW_VCF_SITES        } from '../../../modules/nf-core/bcftools/view/main.nf'
+include { BCFTOOLS_INDEX                         } from '../../../modules/nf-core/bcftools/index/main.nf'
+include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2     } from '../../../modules/nf-core/bcftools/index/main.nf'
```
Comment (LouisLeNezet) [inline subworkflows/local/vcf_sites_extract_bcftools/main.nf:4]: We should stick to one method to rename modules when we do multiple import.
Which one would you prefer INDEX or BCFTOOLS_INDEX ?

File: subworkflows/local/vcf_phase_panel/main.nf [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        Channel.of([[],[],[]]).collect(),
+        Channel.of([[],[]]).collect())
+        ch_versions = ch_versions.mix(VCF_PHASE_SHAPEIT5.out.versions)
+        ch_panel_phased = VCF_PHASE_SHAPEIT5.out.variants_phased
+            .combine(VCF_PHASE_SHAPEIT5.out.variants_index, by: 0)
+    } else {
+        ch_panel_phased = ch_vcf
+    }
+
+    ch_panel = ch_panel_norm
```
Comment (LouisLeNezet) [inline subworkflows/local/vcf_phase_panel/main.nf:28]: I think this combination of channels should be made in the main script.
This subworkflow should only phase the panel (maybe the if statement should also be in the main script).

File: workflows/phaseimpute/main.nf (hunk 7 above)
Comment (LouisLeNezet) [inline workflows/phaseimpute/main.nf:117]: Great job to reorganize all that 😄 
If it become too long we could put all that in a dedicated subworkflow to not have too much lines in the main script.
What do you think ?

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf (hunk 5 above)
Comment (LouisLeNezet) [reply subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:23]: Same for the bcftools query ?

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    // Get chromosomes of posfile
+    ch_posfile = GAWK.out.output.map{meta, posfile -> return[['chr': meta.chr], posfile]}
+
+    // Get chromosomes of fasta
+    ch_chromosomes = ch_fasta.map{it -> it[2]}
```
Comment (LouisLeNezet) [inline subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:29]: If I'm not wrong this chromosome extraction is done on the initialisation level.
We could add a channel as input to link it ?

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                    .map{it -> return [[chr: it.chr], it.chr]}
+
+    // Combine channels
+    def input_empty         = [[]]
+    def rdata_empty         = [[]]
+    k_val                   = params.k_val
+    ngen                    = params.ngen
+
+    // Make final channel with parameters
+    stitch_parameters = ch_posfile.map { it + input_empty + rdata_empty}
```
Comment (LouisLeNezet) [inline subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:40]: If I understand well this is also pre-processing of the panel ?

File: docs/usage.md [hunk 12]
```diff
@@ -135,14 +127,26 @@ You can also generate such `YAML`/`JSON` files via [nf-core/launch](https://nf-c
...
+nextflow run nf-core/phaseimpute --input samplesheet.csv --panel samplesheet_reference.csv --step impute --tool quilt --outdir results --genome GRCh37 -profile docker
+```
+
+#### STITCH
+
+```bash
+nextflow run nf-core/phaseimpute --input samplesheet.csv --step impute --tool stitch --outdir results --genome GRCh37 -profile docker
+```
+
+Notice that no reference panel is needed when running STITCH.
```
Comment (LouisLeNezet) [inline docs/usage.md:144]: You said that a reference panel is not needed but the preparation of the parameters require:
"VCF_SITES_EXTRACT_BCFTOOLS.out.panel_sites". How can we manage that ?

File: conf/test_stitch.config [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    max_time   = '1.h'
+
+    // Input data
+    input        = "${projectDir}/tests/csv/sample_bam.csv"
+    input_region = "${projectDir}/tests/csv/region.csv"
+
+    // Genome references
+    fasta  = "https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa"
+    phased = true
+    panel  = "${projectDir}/tests/csv/panel.csv"
```
Comment (LouisLeNezet) [inline conf/test_stitch.config:29]: Shouldn't we remove the panel as stitch normally doesn't need it ?

**Review by LouisLeNezet — CHANGES_REQUESTED** [review]

**Comment by LouisLeNezet** [comment]
Great job @atrigila !
We should also add the test for stitch in the github actions 😉

File: docs/usage.md (hunk 12 above)
Comment (atrigila) [reply docs/usage.md:144]: There should be a `--posfile` parameter where users can input their own sites. This means that when running `--mode impute`, it should not run the `panelprep` mode (it is currently doing that for all tools by default).

If a tool needs the `panelprep` mode, it would be better to specify that by using `--mode panelprep,impute`.

File: conf/test_stitch.config (hunk 13 above)
Comment (atrigila) [reply conf/test_stitch.config:29]: When removing, it currently breaks because prepare panel includes `params.step == 'impute'`. This should be removed (but I'll have to modify `glimpse1` as well). 

```

    //
    // Prepare panel
    //
    if (params.step == 'impute' || params.step == 'panel_prep' || params.step == 'validate' || params.step == 'all') {
        // Check chr prefix and remove if necessary

```

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf (hunk 11 above)
Comment (atrigila) [reply subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:40]: This is not preprocessing, but part of the input channel preparation for stitch.

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf (hunk 10 above)
Comment (atrigila) [reply subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:29]: Good idea, I'll check this!

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf (hunk 8 above)
Comment (atrigila) [reply subworkflows/local/vcf_sites_extract_bcftools/main.nf:4]: ` BCFTOOLS_INDEX` is more informative

File: nextflow_schema.json (hunk 2 above)
Comment (atrigila) [reply nextflow_schema.json:76]: I think this was the linter, thanks for spotting it!

File: workflows/phaseimpute/main.nf (hunk 7 above)
Comment (atrigila) [reply workflows/phaseimpute/main.nf:117]: I agree!

File: subworkflows/local/vcf_phase_panel/main.nf (hunk 9 above)
Comment (atrigila) [reply subworkflows/local/vcf_phase_panel/main.nf:28]: Mmm, the issue I see is that if the subworkflow is only to phase the panel, then it would be the same as `VCF_PHASE_SHAPEIT5`.

File: conf/steps/panel_prep.config [hunk 14]
```diff
@@ -25,27 +25,41 @@ process {
...
-        publishDir = [
-            path: { "${params.outdir}/prep_panel/" },
-            mode: params.publish_dir_mode,
-            enabled: true,
-            saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
-        ]
+        withName: 'NFCORE_PHASEIMPUTE:PHASEIMPUTE:VCF_NORMALIZE_BCFTOOLS:BCFTOOLS_NORM' {
+        ext.args      = '-m +any --no-version --output-type z'
+        ext.prefix    = { "${meta.id}_${meta.chr}_multiallelic" }
+        cpus          = 2
```
Comment (LouisLeNezet) [inline conf/steps/panel_prep.config:34]: Why do you add the cpu, memory and maxRetries each time ?
I'm not sure this is necessary 🤔

File: conf/test_stitch.config [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    max_memory = '2.GB'
+    max_time   = '1.h'
+
+    // Input data
+    input        = "${projectDir}/tests/csv/sample_bam.csv"
+    input_region = "${projectDir}/tests/csv/region.csv"
+
+    // Genome references
+    fasta  = "https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa"
+    phased = true
```
Comment (LouisLeNezet) [inline conf/test_stitch.config:28]: I don't think you need the phased = true.

File: docs/output.md [hunk 16]
```diff
@@ -21,40 +19,60 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
 - [Convert](#convert) - Convert reference panel to .hap and .legend files
 - [QUILT](#quilt) - Perform imputation
 - [Concatenate](#concatenate) - Concatenate all imputed chunks into a single VCF.
-- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
```
Comment (LouisLeNezet) [inline docs/output.md:24]: We still have the pipeline and mutliqc steps.

File: docs/usage.md [hunk 17]
```diff
@@ -133,16 +125,36 @@ genome: 'GRCh37'
 
 You can also generate such `YAML`/`JSON` files via [nf-core/launch](https://nf-co.re/launch).
 
-### Imputation modes
+### Imputation tools `--mode impute --tools`
```
Comment (LouisLeNezet) [inline docs/usage.md:128]: This should be --step instead of --mode

File: modules/nf-core/stitch/stitch.diff (hunk 1 above)
Comment (LouisLeNezet) [reply modules/nf-core/stitch/stitch.diff:10]: I think we could add that directly to nf-core module.
What do you think ?

File: nextflow_schema.json [hunk 18]
```diff
@@ -73,81 +136,18 @@
...
                     "type": "number",
-                    "description": "Minimum genotype likelihood probability P(G|R) in validation data. Set to zero to have no filter of if using –gt-validation",
+                    "description": "Minimum genotype likelihood probability P(G|R) in validation data. Set to zero to have no filter of if using gt-validation",
                     "default": 0.9,
                     "pattern": "^\\d+(\\.\\d+)?$"
                 },
                 "min_val_dp": {
                     "type": "integer",
-                    "description": "Minimum coverage in validation data. If FORMAT/DP is missing and –min_val_dp > 0, the program exits with an error. Set to zero to have no filter of if using –gt-validation",
+                    "description": "Minimum coverage in validation data. If FORMAT/DP is missing and \u2013min_val_dp > 0, the program exits with an error. Set to zero to have no filter of if using \u2013gt-validation",
```
Comment (LouisLeNezet) [inline nextflow_schema.json:145]: Why did you change to \u2013 instead of a minus ?

File: subworkflows/local/vcf_concatenate_bcftools/main.nf [hunk 19]
```diff
@@ -11,8 +11,7 @@ workflow VCF_CONCATENATE_BCFTOOLS {
     ch_versions = Channel.empty()
 
     // Remove chromosome from meta
-    ch_vcf_tbi_grouped = ch_vcf_tbi
-        .map{ meta, vcf, tbi -> [['id' : meta.id], vcf, tbi] }
+    ch_vcf_tbi_grouped = ch_vcf_tbi.map{ meta, vcf, tbi -> [['id' : meta.id], vcf, tbi] }
 
     // Group by ID
     ch_vcf_tbi_grouped = ch_vcf_tbi_grouped.groupTuple( by:0 )
```
Comment (LouisLeNezet) [inline subworkflows/local/vcf_concatenate_bcftools/main.nf:17]: I'm not sure if the grouping needs to be done inside this subworkflow or outside when we call it.
What do you think ?

File: subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    } else if (params.posfile){
+        // Get unique chromosomes in posfile
+        ch_chromosomes_posfile = ch_posfile.map{it -> it[1]}
+                .splitCsv(header: ["chr", "pos", "ref", "alt"], sep: "\t")
+                .map{it -> return [it.chr]}
+
+        ch_chromosomes_posfile = ch_chromosomes_posfile.unique()
+
+        // Make final channel with parameters
+        stitch_parameters = ch_posfile.map { it + input_empty + rdata_empty}
```
Comment (LouisLeNezet) [inline subworkflows/local/prepare_input_stitch/prepare_input_stitch.nf:40]: Isn't ch_posfile already separated by chromosomes ?
If so it should have the "chr" inside its meta map, not ?

File: subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    main:
+
+    ch_versions      = Channel.empty()
+
+    // Convert position file to tab-separated file
+    BCFTOOLS_QUERY(ch_panel_sites, [], [], [])
+    ch_posfile = BCFTOOLS_QUERY.out.output
+
+    // Remove multiallelic positions from tsv
+    GAWK(ch_posfile, [])
```
Comment (LouisLeNezet) [inline subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf:20]: I don't know if it's better or not to give the script in the config file directly or in a file with collectFile()...
We might need to normalize this usage with GAWK.

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf (hunk 8 above)
Comment (LouisLeNezet) [reply subworkflows/local/vcf_sites_extract_bcftools/main.nf:4]: Okay so we might need to change it in the rest of the pipeline.
I'll add it to the issues ;)

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+    BCFTOOLS_INDEX_2(VIEW_VCF_SITES.out.vcf)
+    ch_versions = ch_versions.mix(BCFTOOLS_INDEX_2.out.versions.first())
+
+    // Join extracted sites and index
+    ch_panel_sites = VIEW_VCF_SITES.out.vcf.combine(BCFTOOLS_INDEX_2.out.csi, by:0)
+
+    // Create TSVs for different tools
+
+    // Convert to TSV with structure for Glimpse
+    BCFTOOLS_QUERY(ch_panel_sites, [], [], [])
```
Comment (LouisLeNezet) [inline subworkflows/local/vcf_sites_extract_bcftools/main.nf:46]: I think that this tsv file is the same as the one used by stitch.
We might need to compare it to see if we can simplify this part.

**Review by LouisLeNezet — CHANGES_REQUESTED** [review]

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf (hunk 22 above)
Comment (atrigila) [reply subworkflows/local/vcf_sites_extract_bcftools/main.nf:46]: It is a different TSV. As far as I saw from the output files, glimpse separates the last two "columns" with a comma, while stitch uses a tab.

File: subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf (hunk 21 above)
Comment (atrigila) [reply subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf:20]: Sarek also gives the script in the config file directly (example here: https://github.com/search?q=repo%3Anf-core%2Fsarek%20BUILD_INTERVALS&type=code). I've found this method more straightforward.

File: docs/output.md (hunk 16 above)
Comment (atrigila) [reply docs/output.md:24]: Yes, it is now in a different section of the `output.md`. It is in the section `##Reports`

File: subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf (hunk 21 above)
Comment (LouisLeNezet) [reply subworkflows/local/prepare_input_stitch/prepare_posfile_tsv.nf:20]: Okay, so I'll update that for the other tools.
I've add that as an issue.

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf (hunk 22 above)
Comment (LouisLeNezet) [reply subworkflows/local/vcf_sites_extract_bcftools/main.nf:46]: That's unfortunate... they could have used the same file 😞

File: docs/output.md (hunk 16 above)
Comment (LouisLeNezet) [reply docs/output.md:24]: Ah, sorry didn't saw it ^^

File: subworkflows/local/vcf_concatenate_bcftools/main.nf (hunk 19 above)
Comment (atrigila) [reply subworkflows/local/vcf_concatenate_bcftools/main.nf:17]: I prefer them inside subworkflows, looks tidier, but I am not sure if there is a recommendation/best-practice around this. Sarek also seems to do these type of channel manipulation inside the local subworkflows ([example here](https://github.com/nf-core/sarek/blob/ea88402912c329b4fd234ad07294ce05bbd2590c/subworkflows/local/bam_applybqsr/main.nf))

File: .nf-core.yml [hunk 23]
```diff
@@ -1 +1,3 @@
 repository_type: pipeline
+lint:
+  subworkflow_changes: false
```
[the remark this answers is not in the document]
Comment (atrigila) [reply .nf-core.yml:3]: As there is no `nf-core subworkflows patch` yet, any changes made to a subworkflow will appear as errors in the linter. This code avoids tracking changes until `nf-core subworkflows patch` is developed.
